### PR TITLE
Feature/regressor improvements

### PR DESCRIPTION
**Pull request by james-p-foster**
If merged this PR will add methods to `JointTorqueRegressorCalculator` which allows partial computation of the regressor matrix, both on a per-rigid-body basis and a per-parameter basis.

### CODE REVIEW

File: src/main/java/us/ihmc/mecano/algorithms/JointTorqueRegressorCalculator.java [hunk 1]
```diff
@@ -3,19 +3,13 @@
 import org.ejml.data.DMatrix;
 import org.ejml.data.DMatrixRMaj;
 import org.ejml.dense.row.CommonOps_DDRM;
-import us.ihmc.mecano.multiBodySystem.interfaces.JointReadOnly;
-import us.ihmc.mecano.multiBodySystem.interfaces.MultiBodySystemBasics;
-import us.ihmc.mecano.multiBodySystem.interfaces.MultiBodySystemReadOnly;
-import us.ihmc.mecano.multiBodySystem.interfaces.RigidBodyBasics;
+import us.ihmc.mecano.multiBodySystem.interfaces.*;
```
Comment (SylvainBertrand) [inline src/main/java/us/ihmc/mecano/algorithms/JointTorqueRegressorCalculator.java:6]: No need to do anything in this PR, but don't forget to update your intellij java code style to the most recent one

File: src/main/java/us/ihmc/mecano/algorithms/JointTorqueRegressorCalculator.java [hunk 2]
```diff
@@ -85,6 +79,11 @@ public class JointTorqueRegressorCalculator
     */
    private final JointTorqueRegressorRecursionStep[] jointTorqueRegressorRecursionSteps;
 
+   /**
+    * If computing blocks of the joint torque regressor matrix for individual bodies, this map can be used to perform body-wise computations.
+    */
+   private final Map<RigidBodyReadOnly, JointTorqueRegressorRecursionStep> rigidBodyToRecursionStepMap = new LinkedHashMap<>();
```
Comment (SylvainBertrand) [inline src/main/java/us/ihmc/mecano/algorithms/JointTorqueRegressorCalculator.java]: Heads up here: `LinkedHashMap` allow to get a map that maintains the order in which the elements were added. This comes at a slight performance penalty when compared to the `HashMap` implementation. If you don't care about the ordering in the map, I'd recommend using a simple `HashMap`. Worth noting that in this case, I'm doubtful that you'd see a noticeable difference in speed.

File: src/main/java/us/ihmc/mecano/algorithms/JointTorqueRegressorCalculator.java [hunk 3]
```diff
@@ -191,11 +189,125 @@ public void compute()
...
+   {
+      // Set all spatial inertias to zero
+      initialRecursionStep.setSpatialInertiasToZeroRecursively();
+
+      // Perform an initial forward pass of the inverse dynamics calculator (which roughly corresponds to the forward
+      // kinematics)
+      inverseDynamicsCalculator.initializeJointAccelerationMatrix(null);
+      inverseDynamicsCalculator.initialRecursionStep.passOneRecursive();
+
+      rigidBodyToRecursionStepMap.get(body).markUpstreamAsModifiedRecursively();
```
Comment (SylvainBertrand) [inline src/main/java/us/ihmc/mecano/algorithms/JointTorqueRegressorCalculator.java]: Keep in mind that retrieving a value in a map is not as trivial as accessing the i-th element of an array. If you need to interact multiple times with the same value, I'd suggest storing it in a local variable so you perform only 1 lookup.
Out of curiosity, just take a peek at the algorithm to retrieve a value in a `HashMap`: https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/util/HashMap.java#L565

File: src/main/java/us/ihmc/mecano/algorithms/JointTorqueRegressorCalculator.java [hunk 4]
```diff
@@ -279,6 +391,40 @@ public DMatrixRMaj getParameterVector()
...
+    * <p>
+    * For more details, see {@link JointTorqueRegressorRecursionStep#parameterVector}.
+    * </p>
+    *
+    * @param body the rigid body to get the inertial parameter vector slice for.
+    * @return the 10-by-1 vector of inertial parameters.
+    */
+   public DMatrixRMaj getParameterVectorSlice(RigidBodyReadOnly body)
+   {
+      return rigidBodyToRecursionStepMap.get(body).parameterVector;
```
Comment (SylvainBertrand) [inline src/main/java/us/ihmc/mecano/algorithms/JointTorqueRegressorCalculator.java]: Let's add some null check like in: https://github.com/ihmcrobotics/mecano/blob/develop/src/main/java/us/ihmc/mecano/algorithms/InverseDynamicsCalculator.java#L578-L584

**Review by SylvainBertrand — CHANGES_REQUESTED** [review]
This is sweet man! Make sure your tests capture silly bugs by mutating your code in couple places and checking that the test fails. Other than that, I just have couple comments.

File: src/main/java/us/ihmc/mecano/algorithms/JointTorqueRegressorCalculator.java (hunk 1 above)
Comment (james-p-foster) [reply src/main/java/us/ihmc/mecano/algorithms/JointTorqueRegressorCalculator.java:6]: 👍🏻 have done.

File: src/main/java/us/ihmc/mecano/algorithms/JointTorqueRegressorCalculator.java (hunk 2 above)
Comment (james-p-foster) [reply src/main/java/us/ihmc/mecano/algorithms/JointTorqueRegressorCalculator.java]: Yeah you're right. But, start as I mean to go on, I've made the change.

File: src/main/java/us/ihmc/mecano/algorithms/JointTorqueRegressorCalculator.java (hunk 3 above)
Comment (james-p-foster) [reply src/main/java/us/ihmc/mecano/algorithms/JointTorqueRegressorCalculator.java]: That's awesome stuff, done.

File: src/main/java/us/ihmc/mecano/algorithms/JointTorqueRegressorCalculator.java (hunk 4 above)
Comment (james-p-foster) [reply src/main/java/us/ihmc/mecano/algorithms/JointTorqueRegressorCalculator.java]: Done a null check on the recursion step, but shouldn't this in my case be a null check on the inertia of the rigid body? I'm unsure. Also, I tried to add a similar null check to `getParameter` below, but that returns a double, so I can only return a `Double.NaN` in that case. I'm not sure that's good?

**Review by SylvainBertrand — APPROVED** [review]
That looks good to me!

**Pull request merged**